### PR TITLE
[Draft] Create SecretGenControllerConfig API and controller

### PR DESCRIPTION
**Pull request by blc1996**
### Which issue(s) this PR fixes

Create SecretGenControllerConfig API and controller

Fixes 

### Describe testing done for PR
Implemented integration tests in local simulated cluster

### Release note

```release-note
Create SecretGenControllerConfig API and controller
```

### PR Checklist

- [ ] Squash the commits into one or a small number of logical commits
      
- [ ] Use good commit [messages](https://github.com/vmware-tanzu/tanzu-framework/blob/main/CONTRIBUTING.md)
- [ ] Ensure PR contains terms all contributors can understand and links all contributors can access

### CODE REVIEW

File: addons/controllers/secretgen-controller/secretgencontrollerconfig_controller.go [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+			log.Info("Cluster not found")
+			return ctrl.Result{}, nil
+		}
+
+		log.Error(err, "unable to fetch cluster")
+		return ctrl.Result{}, err
+	}
+
+	// Get client for the cluster that we deploy data value secret
+	remoteClient, err := util.GetClusterClient(ctx, r.Client, r.Scheme, clusterapiutil.ObjectKey(cluster))
```
Comment (maralavi) [inline addons/controllers/secretgen-controller/secretgencontrollerconfig_controller.go:78]: Wh do we need to come up with ```remoteClient```? Is this same controller going to eventually workload clusters config too? I mean why not using ```r.client``` itself?

File: addons/controllers/secretgen-controller/secretgencontrollerconfig_controller.go [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+	// get secret-gen config object
+	secretGenConfig := &addonsv1alpha1.SecretGenControllerConfig{}
+	if err := r.Client.Get(ctx, req.NamespacedName, secretGenConfig); err != nil {
+		log.Error(err, "unable to fetch SecretGenControllerConfig")
+		return ctrl.Result{}, err
+	}
+
+	// verify that the cluster related to config is present
+	cluster := &clusterapiv1beta1.Cluster{}
+	if err := r.Client.Get(ctx, req.NamespacedName, cluster); err != nil {
```
Comment (maralavi) [inline addons/controllers/secretgen-controller/secretgencontrollerconfig_controller.go:67]: Does the ```req.NamespacedName``` really involve cluster name and namespace? Did you test this on an existing cluster?

File: addons/controllers/secretgen-controller/secretgencontrollerconfig_controller.go [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+		if apierrors.IsNotFound(err) {
+			log.Info("Cluster not found")
+			return ctrl.Result{}, nil
+		}
+
+		log.Error(err, "unable to fetch cluster")
+		return ctrl.Result{}, err
```
Comment (maralavi) [inline addons/controllers/secretgen-controller/secretgencontrollerconfig_controller.go:74]: Could just in a single line do the following: 
```
return ctrl.Result{}, client.IgnoreNotFound(err)
```

File: addons/controllers/secretgen-controller/secretgencontrollerconfig_controller.go [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	clusterapiv1beta1 "sigs.k8s.io/cluster-api/api/v1beta1"
+
+	"github.com/go-logr/logr"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	addonsv1alpha1 "github.com/vmware-tanzu/tanzu-framework/apis/addons/v1alpha1"
```
Comment (danniel1205) [inline addons/controllers/secretgen-controller/secretgencontrollerconfig_controller.go:34]: nit: linter might complain on the imports ordering.

File: addons/controllers/secretgen-controller/secretgencontrollerconfig_controller.go [hunk 5]
```diff
@@ -0,0 +1,238 @@
...
+	if !clusterapiutil.HasOwnerRef(secretGenConfig.OwnerReferences, ownerReference) {
+		log.Info("Adding owner reference to SecretGenControllerConfig")
+		secretGenConfig.OwnerReferences = clusterapiutil.EnsureOwnerRef(secretGenConfig.OwnerReferences, ownerReference)
+		*patchCRD = true
+	}
+
+	addonDataValuesSecret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      util.GenerateDataValueSecretNameFromAddonNames(secretGenConfig.Name, constants.SecretGenControllerAddonName),
+			Namespace: secretGenConfig.Namespace,
```
Comment (maralavi) [inline addons/controllers/secretgen-controller/secretgencontrollerconfig_controller.go:177]: Does ```secretGenConfig.Name``` matches the name of the cluster. Why that is the case? (I ask cause the first parameter of the ```GenerateDataValueSecretNameFromAddonNames``` is named as cluster name).

File: addons/controllers/secretgen-controller/secretgencontrollerconfig_controller.go (hunk 2 above)
Comment (danniel1205) [reply addons/controllers/secretgen-controller/secretgencontrollerconfig_controller.go:67]: Question: Using `req.NamespacedName` for both `xxxConfig` and `Cluster` implicitly assumes two objs always have the same name and under the same namespace. We need to be very careful even we are opinionated on naming `xxxConfig`.

File: addons/controllers/secretgen-controller/secretgencontrollerconfig_controller.go (hunk 1 above)
Comment (danniel1205) [reply addons/controllers/secretgen-controller/secretgencontrollerconfig_controller.go:78]: Question: `xxxConfig` API exists on mgmt cluster only, using `r.Client` should be sufficient.

File: addons/controllers/secretgen-controller/secretgencontrollerconfig_controller.go [hunk 6]
```diff
@@ -0,0 +1,238 @@
...
+
+		// marshall the yaml contents
+		secretGenConfigYaml := &SecretGenControllerConfigSpecYaml{Namespace: secretGenConfig.Spec.Namespace, CreateNamespace: secretGenConfig.Spec.CreateNamespace}
+
+		yamlBytes, err := yaml.Marshal(secretGenConfigYaml)
+		if err != nil {
+			return err
+		}
+
+		dataValueBytes := append([]byte(constants.TKGDataValueFormatString), yamlBytes...)
```
Comment (maralavi) [inline addons/controllers/secretgen-controller/secretgencontrollerconfig_controller.go:193]: Do we need YTT annotations for or case in the header or just key valued configuration values wwould be enough?

File: addons/controllers/secretgen-controller/secretgencontrollerconfig_controller.go (hunk 5 above)
Comment (blc1996) [reply addons/controllers/secretgen-controller/secretgencontrollerconfig_controller.go:177]: secretGenConfig will have the same namespaced name as the cluster it relates to.

File: PROJECT [hunk 7]
```diff
@@ -23,4 +23,7 @@ resources:
 - group: cni
```
[the remark this answers is not in the document]
Comment (blc1996) [reply PROJECT:23]: Yes, that's also my take. @maralavi

File: addons/controllers/secretgen-controller/secretgencontrollerconfig_controller.go (hunk 1 above)
Comment (blc1996) [reply addons/controllers/secretgen-controller/secretgencontrollerconfig_controller.go:78]: We need to deploy the data value secret into the workload clusters (remote cluster)

File: addons/controllers/secretgen-controller/secretgencontrollerconfig_controller.go [hunk 8]
```diff
@@ -0,0 +1,238 @@
...
+	// Add finalizer to addon secret
+	*patchCRD = util.AddFinalizerToCRD(log, constants.SecretGenControllerAddonName, secretGenConfig)
+
+	// add owner reference to addon secret
+	ownerReference := metav1.OwnerReference{
+		APIVersion:         clusterapiv1beta1.GroupVersion.String(),
+		Kind:               "Cluster",
+		Name:               cluster.Name,
+		UID:                cluster.UID,
+		Controller:         pointer.BoolPtr(true),
```
Comment (danniel1205) [inline addons/controllers/secretgen-controller/secretgencontrollerconfig_controller.go:164]: question: any particular reason we want to add `controller:true` ? Are we going to enqueue cluster on `xxxConfig` deletion for garbage collection ?

**Comment by blc1996** [comment]
> @vijaykatam @blc1996 Do we need to do all of the patch helper, checking DeletionTimestamp and having different Normal and Delete reconcilers for a Config object? Don't we only want to update the status of the CalicoConfig object?

We need to add finalizer to the config CRD because we need to delete the corresponding data value secret when the CRD is deleted.

File: addons/controllers/secretgen-controller/secretgencontrollerconfig_controller.go [hunk 9]
```diff
@@ -0,0 +1,238 @@
...
+		if apierrors.IsNotFound(err) {
+			log.Info("SecretGenControllerConfig data values secret not found")
+			return nil
+		}
+		log.Error(err, "Error deleting SecretGenControllerConfig data values secret")
+		return err
+	}
+
+	// Remove finalizer from addon secret
+	*patchCRD = util.RemoveFinalizerFromCRD(log, constants.SecretGenControllerAddonName, secretGenConfig)
```
Comment (maralavi) [inline addons/controllers/secretgen-controller/secretgencontrollerconfig_controller.go:235]: Why do we need to patch CRD eventhough it is just getting deleted?

File: PROJECT (hunk 7 above)
[the remark this answers is not in the document]
Comment (maralavi) [reply PROJECT:23]: I'm fine to change the group from cni to addons. cc: @vijaykatam

File: addons/controllers/secretgen-controller/suite_test.go [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+		}
+		for _, crdDir := range crdDirs {
+			crdPaths = append(crdPaths, filepath.Join(strings.TrimSuffix(string(gopath), "\n"),
+				"pkg", "mod", strings.TrimSuffix(string(depPath), "\n"), crdDir))
+		}
+	}
+
+	logf.Log.Info("external CRD paths", "crdPaths", crdPaths)
+	return crdPaths, nil
+}
```
Comment (maralavi) [inline addons/controllers/secretgen-controller/suite_test.go:162]: If we move up the controllers for Config API objects to the same level as the addons controller, we don't need to rewrite the same suite_test.go over all subdirectories. Maybe something to consider

File: addons/pkg/util/addon_util.go [hunk 11]
```diff
@@ -70,6 +73,11 @@ func GenerateAppSecretNameFromAddonSecret(addonSecret *corev1.Secret) string {
 	return fmt.Sprintf("%s-data-values", GenerateAppNameFromAddonSecret(addonSecret))
 }
 
+// GenerateDataValueSecretNameFromAddonAndClusterNames generates data value secret name from addon names
```
Comment (shivaani0505) [inline addons/pkg/util/addon_util.go:76]: nit: I think you were intending to use the name in comment for function name.Semantically it sounds better to me.
You can choose to update comment or function name.

File: addons/pkg/constants/constants.go [hunk 12]
```diff
@@ -97,4 +97,10 @@ const (
 
 	// CRDWaitPollTimeout is poll timeout for checking server resources
 	CRDWaitPollTimeout = time.Minute * 10
+
+	// DataValueFileName is the name of data value file
+	DataValueFileName = "values.yaml"
+
+	// SecretGenControllerAddonName is the name of sectre-gen-controller
```
Comment (shivaani0505) [inline addons/pkg/constants/constants.go:104]: nit: Change to sectre-gen-controller => secretgen-controller

File: addons/controllers/secretgen-controller/secretgencontrollerconfig_controller.go [hunk 13]
```diff
@@ -0,0 +1,238 @@
...
+	if retResult, err := r.ReconcileSecretGenControllerConfig(ctx, log, cluster, remoteClient, secretGenConfig); err != nil {
+		log.Error(err, "unable to reconcile SecretGenControllerConfig")
+		return retResult, err
+	}
+
+	return ctrl.Result{}, nil
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *SecretGenControllerConfigReconciler) SetupWithManager(ctx context.Context, mgr ctrl.Manager, options controller.Options) error {
```
Comment (shivaani0505) [inline addons/controllers/secretgen-controller/secretgencontrollerconfig_controller.go:93]: mgr and options are unused arguments in SetupWithManager function

File: addons/main.go [hunk 14]
```diff
@@ -26,6 +26,7 @@ import (
 	kapppkg "github.com/vmware-tanzu/carvel-kapp-controller/pkg/apis/packaging/v1alpha1"
 	kappdatapkg "github.com/vmware-tanzu/carvel-kapp-controller/pkg/apiserver/apis/datapackaging/v1alpha1"
 	"github.com/vmware-tanzu/tanzu-framework/addons/controllers"
+	secretGen "github.com/vmware-tanzu/tanzu-framework/addons/controllers/secretGen-controller"
```
Comment (shivaani0505) [inline addons/main.go:29]: In this import, change secret**G**en-controller => secretgen-controller

File: addons/controllers/secretgen-controller/secretgencontrollerconfig_controller.go (hunk 1 above)
Comment (shyaamsn) [reply addons/controllers/secretgen-controller/secretgencontrollerconfig_controller.go:78]: AFAIK the secret has to be created in the management/supervisor cluster and xxConfig.status.secretRef set to the secret name. 

TanzuBootStrapCluster controller will copy the secret to workload cluster.

@vijaykatam  @danniel1205  to confirm

Comment (vijaykatam) [reply addons/controllers/secretgen-controller/secretgencontrollerconfig_controller.go:78]: That is correct, the controller for the individual API types is responsible for creating a secret and adding it to status.

File: PROJECT (hunk 7 above)
[the remark this answers is not in the document]
Comment (vijaykatam) [reply PROJECT:23]: no we decided antrea needs to be in cni group to delineate responsibility for owning the type

File: addons/controllers/secretgen-controller/secretgencontrollerconfig_controller.go (hunk 9 above)
Comment (blc1996) [reply addons/controllers/secretgen-controller/secretgencontrollerconfig_controller.go:235]: If the finalizer is not removed from secretGenConfig, it won't be deleted.

Comment (maralavi) [reply addons/controllers/secretgen-controller/secretgencontrollerconfig_controller.go:235]: True, convinced

**Comment by blc1996** [comment]
Closing as secretgen-controller is not a core package anymore

**Pull request closed without merging**